### DISCUSSION TITLE
How to detect a closed JetStream (due to server restart)?

### DISCUSSION THREAD

**Discussion by barafael** [nats-io/nats.rs#1290] [Q&A]
We are using a `WorkQueue` consumer with `StorageType::Memory` and `AckPolicy::Explicit`. When the server restarts, it forgets the streams which existed before the shutdown. In a consumer, how can we observe that the server has shut down, so we can eventually recreate the stream when the server comes up again?

* We tried relying on the `Stream` interface, which states that the stream yields `None` when exhausted, but we do not observe this.
* We can observe the NATS connection state
    - we could lose messages currently queued in the consumer
    - we cannot detect that we have processed all of them, because of the earlier point

We DO observe an error on the stream, around the time of the reconnection:

````
2024-07-30T15:35:11.045783Z  WARN thing::module::consumer: pull consumer failed. error=pull request failed: request failed: missing field `stream_name` at line 1 column 129
````

**Comment by Jarema** [comment]
That looks like a bug. The consumer should recover on its own.
Will take a look.

**Reply by Jarema** [reply]
@barafael can you please share some  snippet how you create and use the consumer?

**Reply by barafael** [reply]
[https://github.com/barafael/njet-stream](https://github.com/barafael/njet-stream) has a complete minimal example.

In three shells:

Step 1: `nats-server -js --user nats --pass nats --addr 127.0.0.1 -p 23561`
Step 2: `cargo run --bin producer`
Step 3: `cargo run --bin consumer`
Observe messages being delivered.
Step 4: Stop nats server
Observe no more messages
Step 5: Restart nats server
Observe still no messages.

This could of course be the intended behaviour, I don't know. One can work around it by having a dedicated tokio task to recreate the NATS consumer when errors come through the stream and forwarding all messages through a common hole. The normal behaviour of a stream is that it yields a None value on close, but in this case here we can't even rely on that.

**Reply by barafael** [reply]
I am currently using a dedicated tokio task which reacts to errors on the jetstream by recreating the consumer. Is it possible I am losing messages when the consumer is being reconstructed?

**Reply by Jarema** [reply]
@barafael If the stream is gone, being in memory stream, after for example server restart, you should not loose messages, as workqueue stream will remove them only after consumer receives ack.

However, I wonder why you are using memory based consumer if this is a persistent consumer?

**Reply by barafael** [reply]
@Jarema apologies for the late reply.
Is memory based but persistent consumer not a legitimate use case? We don't want to pummel the disk.

I do think the stream trait implementation is misbehaving here. It's forcing me to have an extra task just to monitor the connection and might be slightly racy because I have to recreate the consumer.

**Reply by Jarema** [reply]
No worries.
I'm not sure I understand - how you want to achieve R1 in-memory persistence of a consumer?
If you restart the server - every state that was in-memory - is gone. 
You have to either use more replicas (3, but of course if all replicas will die at the same time, you will loose data anyway), or file-based.

What do you think is misbehaving? Please elaborate a bit.

**Reply by barafael** [reply]
Thanks for the insights!

I accept the data loss in case of nats server restart.

Regarding the misbehaviour: see the reproduction above.
Here's how I understand the Stream trait: `next()` (or really, `poll_next()`), yields an item once ready (wrapped in a `Some(.)`). Great. Then we can poll it again. Yields another item. Great. Once it yields `None`, we know the stream is closed and may be dropped. This is how e.g. the `TcpStream` from tokio does it and I think it's the intended behaviour. If an error happens on a stream, I believe it may simply be yielded as well, but doesn't signify the end of stream.

From what I can observe here, once the server was disconnected for whatever reason, the jetstream `Stream` implementation loses the connection (doesn't reconnect). But also it doesn't close. It eventually sends some error though, which we can detect. In that case, we can reopen the connection. But that's overall not that great, I think...

**Reply by Jarema** [reply]
`Stream` does not have any concept of connection by itself.
All it does, is poll core nats subscriptions for messages (and in case of pull requests - also send new requests).
Because of that, it does not have to act in any specific way regarding the connections.
NATS Client subscriptions will be automatically resubsribed when reconnect happens.

And as client is meant to handle reconnecitons under the hood, we do not terminate neither Core NATS subscriptions, nor JetStream Consumers.

We had a concept of terminal errors, but with time we realized, that almost never its a good idea to terminale iterator - as so often things can recover on their own, or get fixed/recreted by other apps, operations etc.

And here is the difference between TCP Stream and NATS -  NATS automatically handle reconnect, so it should not close streams on reconnections.

I'm happy to help you with specific errors you encounter. I would like to learn why you need a separate task.
